File: packages/login-verify/src/oidc/conformance.ts
import {
  ADAPTER_VERSION,
  DIDWEBS_SPEC_VERSION,
  EUDI_ARF_PROFILE_VERSION,
  OPENID4VP_PROFILE,
  REQUIRED_DISCOVERY_FIELDS,
  SD_JWT_VC_PROFILE,
  SIOPV2_PROFILE,
} from "./profile.js";
import type { ConformanceResult, OIDCDiscoveryDocument } from "./types.js";

const VP_FORMAT_SDJWT = "vc+sd-jwt";
const VP_FORMAT_ACDC = "acdc";

/** Validate discovery document against pinned EUDI ARF profile (OWD-2). */
export function validateDiscoveryConformance(doc: OIDCDiscoveryDocument): ConformanceResult {
  const errors: string[] = [];
  const check = (ok: boolean, msg: string) => {
    if (!ok) errors.push(msg);
  };

  for (const field of REQUIRED_DISCOVERY_FIELDS) {
    switch (field) {
      case "issuer":
        check(!!doc.issuer, "missing issuer");
        break;
      case "authorization_endpoint":
        check(!!doc.authorization_endpoint, "missing authorization_endpoint");
        break;
      case "response_types_supported":
        check((doc.response_types_supported?.length ?? 0) > 0, "missing response_types_supported");
        break;
      case "id_token_signing_alg_values_supported":
        check(doc.id_token_signing_alg_values_supported?.includes("EdDSA") ?? false,
          "EdDSA required in id_token_signing_alg_values_supported");
        break;
      case "subject_types_supported":
        check((doc.subject_types_supported?.length ?? 0) > 0, "missing subject_types_supported");
        break;
      case "scopes_supported":
        check(doc.scopes_supported?.includes("openid") ?? false, "openid scope required");
        break;
      case "claims_supported":
        check((doc.claims_supported?.length ?? 0) > 0, "missing claims_supported");
        break;
      case "vp_formats":
        check(!!doc.vp_formats, "missing vp_formats");
        if (doc.vp_formats) {
          check(!!doc.vp_formats[VP_FORMAT_SDJWT], "vp_formats must include vc+sd-jwt");
          check(!!doc.vp_formats[VP_FORMAT_ACDC], "vp_formats must include acdc");
        }
        break;
      case "eudi_arf_profile_version":
        check(doc.eudi_arf_profile_version === EUDI_ARF_PROFILE_VERSION,
          `eudi_arf_profile_version must be ${EUDI_ARF_PROFILE_VERSION}`);
        break;
      case "siopv2_profile":
        check(doc.siopv2_profile === SIOPV2_PROFILE, `siopv2_profile must be ${SIOPV2_PROFILE}`);
        break;
      case "openid4vp_profile":
        check(doc.openid4vp_profile === OPENID4VP_PROFILE, `openid4vp_profile must be ${OPENID4VP_PROFILE}`);
        break;
      case "sd_jwt_vc_profile":
        check(doc.sd_jwt_vc_profile === SD_JWT_VC_PROFILE, `sd_jwt_vc_profile must be ${SD_JWT_VC_PROFILE}`);
        break;
      case "didwebs_spec_version":
        check(doc.didwebs_spec_version === DIDWEBS_SPEC_VERSION,
          `didwebs_spec_version must be ${DIDWEBS_SPEC_VERSION}`);
        break;
      case "ia_oidc_adapter_version":
        check(doc.ia_oidc_adapter_version === ADAPTER_VERSION,
          `ia_oidc_adapter_version must be ${ADAPTER_VERSION}`);
        break;
    }
  }

  return { ok: errors.length === 0, errors, profile: EUDI_ARF_PROFILE_VERSION };
}

File: identity-agent-core/oidc/conformance.go
package oidc

import "fmt"

// ConformanceResult captures EUDI ARF profile validation against pinned versions.
type ConformanceResult struct {
	OK      bool
	Errors  []string
	Profile string
}

// ValidateDiscovery checks a discovery document against the pinned EUDI ARF profile (OWD-2).
func ValidateDiscovery(doc DiscoveryDocument) ConformanceResult {
	res := ConformanceResult{OK: true, Profile: EUDIARFProfileVersion}
	check := func(ok bool, msg string) {
		if !ok {
			res.OK = false
			res.Errors = append(res.Errors, msg)
		}
	}

	for _, field := range RequiredDiscoveryFields {
		switch field {
		case "issuer":
			check(doc.Issuer != "", "missing issuer")
		case "authorization_endpoint":
			check(doc.AuthorizationEndpoint != "", "missing authorization_endpoint")
		case "response_types_supported":
			check(len(doc.ResponseTypesSupported) > 0, "missing response_types_supported")
		case "id_token_signing_alg_values_supported":
			check(containsStr(doc.IDTokenSigningAlgValuesSupported, "EdDSA"), "EdDSA required in id_token_signing_alg_values_supported")
		case "subject_types_supported":
			check(len(doc.SubjectTypesSupported) > 0, "missing subject_types_supported")
		case "scopes_supported":
			check(containsStr(doc.ScopesSupported, "openid"), "openid scope required")
		case "claims_supported":
			check(len(doc.ClaimsSupported) > 0, "missing claims_supported")
		case "vp_formats":
			check(doc.VPFormats != nil, "missing vp_formats")
			if doc.VPFormats != nil {
				check(doc.VPFormats[VPFormatSDJWT] != nil, "vp_formats must include vc+sd-jwt")
				check(doc.VPFormats[VPFormatACDC] != nil, "vp_formats must include acdc")
			}
		case "eudi_arf_profile_version":
			check(doc.EUDIARFProfileVersion == EUDIARFProfileVersion,
				fmt.Sprintf("eudi_arf_profile_version must be %s, got %s", EUDIARFProfileVersion, doc.EUDIARFProfileVersion))
		case "siopv2_profile":
			check(doc.SIOPv2Profile == SIOPv2Profile,
				fmt.Sprintf("siopv2_profile must be %s", SIOPv2Profile))
		case "openid4vp_profile":
			check(doc.OpenID4VPProfile == OpenID4VPProfile,
				fmt.Sprintf("openid4vp_profile must be %s", OpenID4VPProfile))
		case "sd_jwt_vc_profile":
			check(doc.SDJWTVcProfile == SDJWTVcProfile,
				fmt.Sprintf("sd_jwt_vc_profile must be %s", SDJWTVcProfile))
		case "didwebs_spec_version":
			check(doc.DIDWebsSpecVersion == DIDWebsSpecVersion,
				fmt.Sprintf("didwebs_spec_version must be %s", DIDWebsSpecVersion))
		case "ia_oidc_adapter_version":
			check(doc.IAOIDCAdapterVersion == AdapterVersion,
				fmt.Sprintf("ia_oidc_adapter_version must be %s", AdapterVersion))
		}
	}
	return res
}

func containsStr(ss []string, want string) bool {
	for _, s := range ss {
		if s == want {
			return true
		}
	}
	return false
}
